keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,854 @@
+
+One of the main use-cases for this request is applications that have "pools of workers" along with a central system 
+that assigns tasks to one these workers (e.g. [Apache Airflow](https://github.com/apache/airflow) and [Apache Spark](https://github.com/apache/spark)).
+Naturally, if the number of tasks varies over time, the size of the pool should scale with the load.
+However, Deployments/ReplicaSets are not currently application-aware when downscaling, so an active worker may be removed 
+when there was an idle one (leading to wasted work).
+
+As of Kubernetes 1.22, the `controller.kubernetes.io/pod-deletion-cost` annotation from
+[KEP-2255](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2255-pod-cost) is available in BETA.
+However, this approach is not fit for purpose.
+
+Firstly, the `pod-deletion-cost` annotation must be PATCHED on Pods BEFORE the `replicas` count is decreased, leading to the following:
+
+1. it cannot be used with existing resources like [HorizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), 
+   as this would require updating annotations before the autoscaler makes a decision to downscale
+2. to use the feature, developers must create custom-autoscaler that updates annotations BEFORE decreasing `replicas`, 
+   making it inaccessible to all but the most advanced Kubernetes users
+
+Secondly, the `pod-deletion-cost` is an annotation, leading to the following:
+
+1. a PATCH call to kube-apiserver is generated for EACH Pod annotation update, and this is not scalable for large numbers of Pods
+2. annotations are permanent but `pod-deletion-cost` is transient, meaning annotations must be cleared after a scale-down (requiring yet more API calls)
+
+These problems can be addressed if we flip the approach, and allow Deployments/ReplicaSets to "ask" an API which Pods 
+are best to remove when the `replicas` count is decreased.
+
+### Goals
+
+- provide a scalable method for users to influence which Pods are removed when downscaling a Deployment/ReplicaSet
+- improve support for autoscaling in apps that have "pools of workers"
+- prevent a poorly designed (or faulty) `downscalePodPicker` from significantly impacting the Deployment/ReplicaSet behaviour
+- support multiple protocols for the `downscalePodPicker`, starting with REST and expanding to gRPC
+
+### Non-Goals
+
+- guarantee that the Pods "chosen" by the `downscalePodPicker` will be removed (external factors may remove Pods before we do)
+- guarantee that the Pods NOT "chosen" by the `downscalePodPicker` will be spared
+- guarantee that `downscalePodPicker` will only be called once per downscale
+- influence Node-removal decisions of systems like cluster-autoscaler
+- cache results from `downscalePodPicker` (we will call the API every time `replicas` is decreased)
+
+
+## Proposal
+
+- A new `downscalePodPicker` field will be added to `ReplicaSetSpec` and `DeploymentSpec`.
```

Comment:
It probably should be under a `downscalePolicy` field.

Are you still planning to have a comparison order API?

Reply:
> It probably should be under a `downscalePolicy` field.

I am open to placing it under a `downscalePolicy` field. However, I think it might make more sense to keep `downscalePodPicker` __(this KEP)__ and `downscalePolicy` __(a future KEP)__ as separate fields.

In my mind, `downscalePodPicker` will configure HOW the "pod picker" API is accessed, and `downscalePolicy` would determine the overall behavior, which may (or may not) include using the defined `downscalePodPicker`.

For example:

```yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: my-replicaset
spec:
  # (NOT PART OF THIS KEP -- and is a hypothetical)
  downscalePolicy:
    xxxx: xxxx
    podSorting:
      - (Step 1: "pod picker")
      - (Step 2: "pod age")
      - ...
  
  # (Proposed by this KEP)
  downscalePodPicker:
    http:
      host: my-app.my-namespace.svc.cluster.local
      httpHeaders:
        - name: authentication
          valueFrom:
            secretKeyRef:
              key: authentication-token
              name: my-secret
      path: "/downscale-pod-picker"
      port: 443
      scheme: "https"
    maxRetries: 3
    timeoutSeconds: 5

  replicas: 3
  template:
    ...
```

> Are you still planning to have a comparison order API?

That is not part of this KEP, but as above, could be introduced by a future KEP by adding a `downscalePolicy` field.

This KEP only replaces one step of the sorting process with the API.

Reply:
I think it should look like this:

```
downscalePolicy:
  podPicker: ...
  podSorting: ...
```

Reply:
@alculquicondor I like that idea!

But how likely is it that we would implement a `podSorting` feature in the near-to-medium future?
If you think it's not likely, it might just be better to leave it as `downscalePolicy` for simplicity?

Reply:
Always plan for extensibility